src/raftstore/store/util.rs

```diff
@@ -190,6 +191,106 @@ pub fn is_epoch_stale(epoch: &metapb::RegionEpoch, check_epoch: &metapb::RegionE
         || epoch.get_conf_ver() < check_epoch.get_conf_ver()
 }
 
+pub fn check_epoch(
+    req: &RaftCmdRequest,
+    region: &metapb::Region,
+    include_region: bool,
+) -> Result<()> {
+    let (mut check_ver, mut check_conf_ver) = (false, false);
+    if !req.has_admin_request() {
+        // for get/set/delete, we don't care conf_version.
+        check_ver = true;
+    } else {
+        match req.get_admin_request().get_cmd_type() {
+            AdminCmdType::CompactLog
+            | AdminCmdType::InvalidAdmin
+            | AdminCmdType::ComputeHash
+            | AdminCmdType::VerifyHash => {}
+            AdminCmdType::Split => check_ver = true,
+            AdminCmdType::ChangePeer => check_conf_ver = true,
+            AdminCmdType::PrepareMerge
+            | AdminCmdType::CommitMerge
+            | AdminCmdType::RollbackMerge
+            | AdminCmdType::TransferLeader => {
+                check_ver = true;
+                check_conf_ver = true;
+            }
+        };
+    }
+
+    if !check_ver && !check_conf_ver {
+        return Ok(());
+    }
+
+    if !req.get_header().has_region_epoch() {
+        return Err(box_err!("missing epoch!"));
+    }
+
+    let from_epoch = req.get_header().get_region_epoch();
+    let latest_epoch = region.get_region_epoch();
+
+    // should we use not equal here?
+    if (check_conf_ver && from_epoch.get_conf_ver() < latest_epoch.get_conf_ver())
+        || (check_ver && from_epoch.get_version() < latest_epoch.get_version())
+    {
+        debug!(
+            "[region {}] received stale epoch {:?}, mime: {:?}",
+            region.get_id(),
+            from_epoch,
+            latest_epoch
+        );
+        let regions = if include_region {
+            vec![region.to_owned()]
+        } else {
+            vec![]
+        };
+        return Err(Error::StaleEpoch(
+            format!(
+                "latest_epoch of region {} is {:?}, but you \
+                 sent {:?}",
+                region.get_id(),
+                latest_epoch,
+                from_epoch
+            ),
+            regions,
+        ));
+    }
+
+    Ok(())
+}
+
+pub fn check_store_id(req: &RaftCmdRequest, store_id: u64) -> Result<()> {
+    let peer = req.get_header().get_peer();
+    if peer.get_store_id() != store_id {
```

Comment:
Check the ok condition first.

Reply:
any rust style for it? @Hoverbear

Reply:
I agree with @BusyJay myself.

Reply:
The official style guide (https://github.com/rust-lang-nursery/fmt-rfcs/blob/master/guide/expressions.md#control-flow-expressions) does not specifically mention a preference.

Reply:
My consideration here is not about the style, but about the branch miss.

Reply:
Any reference of branch detection about ok condition first?

Reply:
You can bench it yourself. Also you can check the generated asm code from the likely intrinsic.